### PR TITLE
Added postprocessing visualization example.

### PR DESCRIPTION
**Pull request by abhishekchitwar**
Example to showcase the postprocessing with Ansys Flagship Simulations Results. (VTK is generated using pydpf-core).

### CODE REVIEW

File: examples/01-advanced-pyansys-examples/using_meshobject_with_field_data.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+"""
+.. _ref_postprocess_using_meshobjects:
+
+======================================================================
+Use the ``MeshObjectPlot`` class to postprocess Simulation Results
+======================================================================
```
Comment (RobPasMue) [inline examples/01-advanced-pyansys-examples/using_meshobject_with_field_data.py:28]:
```suggestion
====================================================================
Postprocessing simulation results using the ``MeshObjectPlot`` class
====================================================================
```

File: examples/01-advanced-pyansys-examples/using_meshobject_with_field_data.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+The Visualization Interface Tool provides the ``MeshObject`` helper class to relate a custom object.
+With a custom object, you can take advantage of the full potential of the Visualization Interface Tool.
+
+This example shows how to use the ``MeshObjectPlot`` class to plot your custom objects with scalar data on mesh.
+
+"""
+
+###################################################
+# Necessary imports
+# =================================================
```
Comment (RobPasMue) [inline examples/01-advanced-pyansys-examples/using_meshobject_with_field_data.py:39]:
```suggestion
# Necessary imports
# =================
```

File: examples/01-advanced-pyansys-examples/using_meshobject_with_field_data.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+
+import pyvista as pv
+from ansys.tools.visualization_interface.backends.pyvista import PyVistaBackend
+from ansys.tools.visualization_interface import MeshObjectPlot
+from ansys.tools.visualization_interface import Plotter
+from ansys.fluent.core import examples
```
Comment (RobPasMue) [inline examples/01-advanced-pyansys-examples/using_meshobject_with_field_data.py:46]:
```suggestion
from ansys.fluent.core import examples
import pyvista as pv

from ansys.tools.visualization_interface.backends.pyvista import PyVistaBackend
from ansys.tools.visualization_interface import MeshObjectPlot, Plotter
```

File: examples/01-advanced-pyansys-examples/using_meshobject_with_field_data.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+###################################################
+# Download the VTK file
+# =================================================
```
Comment (RobPasMue) [inline examples/01-advanced-pyansys-examples/using_meshobject_with_field_data.py:50]:
```suggestion
# Download the VTK file
# =====================
```

File: examples/01-advanced-pyansys-examples/using_meshobject_with_field_data.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+# A VTK dataset can be produced utilizing the `pydpf <https://dpf.docs.pyansys.com/version/stable/>`_
+# for Ansys Flagship products simulations results file format.
```
Comment (RobPasMue) [inline examples/01-advanced-pyansys-examples/using_meshobject_with_field_data.py:53]:
```suggestion
#
# A VTK dataset can be produced utilizing `PyDPF <https://dpf.docs.pyansys.com/version/stable/>`_
# for Ansys Flagship products simulations results file format.
```

File: examples/01-advanced-pyansys-examples/using_meshobject_with_field_data.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+mixing_elbow_file_src = examples.download_file("mixing_elbow.vtk", "result_files/fluent-mixing_elbow_steady-state")
+
+###################################################
+# Define a custom object class
+# =================================================
+
+# Note that the ``CustomObject`` class must have a way to get the mesh
+# and a name or ID.
```
Comment (RobPasMue) [inline examples/01-advanced-pyansys-examples/using_meshobject_with_field_data.py:62]:
```suggestion
###################################################
# Define a custom object class
# ============================
#
# Note that the ``CustomObject`` class must have a way to get the mesh
# and a name or ID.
```

File: examples/01-advanced-pyansys-examples/using_meshobject_with_field_data.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+######################################
+# Get the available field data arrays
+# ====================================
+
+field_data_arrays = custom_vtk.get_field_array_info()
+print(f"Field data arrays: {field_data_arrays}")
+
+####################################################################
+# Plot the ``MeshObjectPlot`` instance with mesh object & field data
+# ==================================================================
```
Comment (RobPasMue) [inline examples/01-advanced-pyansys-examples/using_meshobject_with_field_data.py:105]:
```suggestion
########################################################################
# Plot the ``MeshObjectPlot`` instance with mesh object & field data (0)
# ======================================================================
```

File: examples/01-advanced-pyansys-examples/using_meshobject_with_field_data.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    show_edges=True,
+    show_scalar_bar=True,
+)
+pl._backend.pv_interface.scene.camera_position = cpos
+pl.show()
+
+
+##########################################################################
+# Plot the ``MeshObjectPlot`` instance with mesh object & other field data
+# ========================================================================
```
Comment (RobPasMue) [inline examples/01-advanced-pyansys-examples/using_meshobject_with_field_data.py:121]:
```suggestion
##############################################################################
# Plot the ``MeshObjectPlot`` instance with mesh object & other field data (1)
# ============================================================================
```

**Review by RobPasMue — CHANGES_REQUESTED** [review]

**Comment by AlejandroFernandezLuces** [comment]
Thanks for adding this example @abhishekchitwar 🙂 

I'm a bit worried about having the fluent dependency in this repo. I believe it might cause dependency issues if we integrate this repo into fluent. Is it possible to include the required example directly in this repository? If not, that's fine, we can always revisit this if PyFluent incorporates this tool.

**Comment by RobPasMue** [comment]
> I'm a bit worried about having the fluent dependency in this repo. I believe it might cause dependency issues if we integrate this repo into fluent. Is it possible to include the required example files directly in this repository? If not, that's fine, we can always revisit this if PyFluent incorporates this tool.

Well, since it is inside the ``docs`` target I wouldn't worry much about it. People will install the package only, or other additional targets. But the ``docs`` target is explicitly for building the documentation of this project.

**Review by RobPasMue — APPROVED** [review]

**Comment by RobPasMue** [comment]
Looks like automerge was activated by @abhishekchitwar - sorry @AlejandroFernandezLuces, should have let it up to you to merge. I didn't see that automerge was active

**Comment by abhishekchitwar** [comment]
@AlejandroFernandezLuces,

Thank you! This issue can be resolved by incorporating a download API in this library to facilitate downloading sample data from the PyAnsys example-data repository. There are no direct dependencies of Fluent apart from downloading example data.

**Comment by AlejandroFernandezLuces** [comment]
> Looks like automerge was activated by @abhishekchitwar - sorry @AlejandroFernandezLuces, should have let it up to you to merge. I didn't see that automerge was active

No worries, it's good to go from my side 🙂

**Comment by AlejandroFernandezLuces** [comment]
> @AlejandroFernandezLuces,
> 
> Thank you! This issue can be resolved by incorporating a download API in this library to facilitate downloading sample data from the PyAnsys example-data repository. There are no direct dependencies of Fluent apart from downloading example data.

We can do that if we have any conflicts with fluent at some point. For now let's keep it simple, thanks gain for adding the example 🚀

**Comment by ahernsean** [comment]
@abhishekchitwar, do note that this dataset has the bug that VTK still considers the mesh cells to be invalid. The values in these cells ought to be zero, not 0x34.
image

**Comment by RobPasMue** [comment]
Is this bug coming from PyVista @ahernsean? Or is it because the VTK file has that invalid info?

**Comment by abhishekchitwar** [comment]
@ahernsean,
I attempted an alternative approach (converting from CGNS to VTK), but I still did not obtain any valid cells. The state described in the VTK documentation is not clear, as it includes:
```
    Valid = 0x0,
    WrongNumberOfPoints = 0x01,
    IntersectingEdges = 0x02,
    IntersectingFaces = 0x04,
    NoncontiguousEdges = 0x08,
    Nonconvex = 0x10,
    FacesAreOrientedIncorrectly = 0x20,
```
The number from the previous snapshot shared by you or the one below suggests that the issue is related to FacesAreOrientedIncorrectly, as we previously discussed. Any suggestions!

image

**Comment by ahernsean** [comment]
@RobPasMue, yes, I believe DPF is likely orienting the faces of polyhedra incorrectly in some cases. @abhishekchitwar and I were working with Paul Profizi on this, who said he'd try a few things and get back to us.

**Comment by ahernsean** [comment]
> The state described in the VTK documentation is not clear, as it includes:
> 
> ```
>     Valid = 0x0,
>     WrongNumberOfPoints = 0x01,
>     IntersectingEdges = 0x02,
>     IntersectingFaces = 0x04,
>     NoncontiguousEdges = 0x08,
>     Nonconvex = 0x10,
>     FacesAreOrientedIncorrectly = 0x20,
> ```

The mapping is clear. Your image suggests that the majority of your cells have a ValidityState of 48 (dec), which is 30 (hex). This corresponds to NoncontiguousEdges and Nonconvex. Precisely what that means about the input mesh is less clear. I would want to restrict down to one or two cells to see what might be happening.

**Pull request merged**